### PR TITLE
[AArch64][ISel] Add support for v8.4a RCpc `ldapur`/`stlur`

### PR DESCRIPTION
**Pull request by antoniofrighetto**
AArch64 backend now features v8.4 Load-Acquire RCpc and Store-Release register unscaled support. As part of the changes, dropping the check for valid scaled immediate in `SelectAddrModeUnscaled` was deemed necessary to achieve this, though if we want to retain this, we might have an helper checking if the uses are `ISD::ATOMIC_LOAD`/`ISD::ATOMIC_STORE` (similar to `isWorthFoldingADDlow`), and bail out only if this is not the case. Minor opportunity to refactor out the check in `isValidAsScaledImmediate` as well.

### CODE REVIEW

**Comment by tschuett** [comment]
Nit. There is no GISel test coverage.

**Comment by antoniofrighetto** [comment]
Right, somehow I thought GISel was the default option with `O0` not exclusively on macOS.

**Comment by tschuett** [comment]
You probably need something like `-global-isel -global-isel-abort=2`.

**Comment by antoniofrighetto** [comment]
Added GISel tests, rebased to main, and dropped unneeded specialized support in GISel for `stlur`, as the execution within the match table will correctly perform the pattern match and replacement (was already happening for `ldapur`).

**Comment by antoniofrighetto** [comment]
Kind ping :) (cc/ @tmatheson-arm, @pratlucas).

**Comment by antoniofrighetto** [comment]
Kind ping

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1130,10 +1139,6 @@ bool AArch64DAGToDAGISel::SelectAddrModeUnscaled(SDValue N, unsigned Size,
     return false;
   if (ConstantSDNode *RHS = dyn_cast<ConstantSDNode>(N.getOperand(1))) {
     int64_t RHSC = RHS->getSExtValue();
-    // If the offset is valid as a scaled immediate, don't match here.
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:1133]: And this doesn't alter any of the existing codegen, for non-atomic types?

**Review by davemgreen — COMMENTED** [review]
I think this looks OK, is the "// If the offset is valid as a scaled immediate, don't match here." part not needed?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 1 above)
Comment (antoniofrighetto) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:1133]: Does not seem to be the case from running current codegen tests, at the very least.

**Comment by antoniofrighetto** [comment]
Not sure about the origin of that addition, though it seems this is no longer required when deciding to leverage unscaled operations (as it hinders their emission). We might introduce a more granular check (as mentioned at the beginning), but this does not look necessary either.

**Review by davemgreen — APPROVED** [review]
Yeah it looks like it will be very rare, if it comes up at all.

LGTM, sorry for the delay in getting the review.

**Comment by antoniofrighetto** [comment]
Thanks a lot for reviewing this :)

**Comment by luporl** [comment]
It looks like this broke some 2-stage AArch64 buildbots:
- https://lab.llvm.org/buildbot/#/builders/198/builds/5636
- https://lab.llvm.org/buildbot/#/builders/176/builds/6627

**Comment by antoniofrighetto** [comment]
> It looks like this broke some 2-stage AArch64 buildbots:
> 
> * https://lab.llvm.org/buildbot/#/builders/198/builds/5636
> * https://lab.llvm.org/buildbot/#/builders/176/builds/6627

Should be fixed by https://github.com/llvm/llvm-project/commit/8ce4b7bcd5f7680243d2faa8eb8340bf92a20907.

**Comment by antoniofrighetto** [comment]
> It looks like this is causing a crash for the following reproducer: https://llvm.godbolt.org/z/brbfGxMKq

Oh oh, we forgot a `LDAPURi`. Thanks for reporting, should be fixed by https://github.com/llvm/llvm-project/commit/c16b94a3bf60d77cb0ea866af416016e672a43d2. At least, comparing the output with `llc-17.0.1`, it seems like we now avoid spilling a few registers + saved some adds. Happy with that!

**Comment by antoniofrighetto** [comment]
@fhahn, sorry, just saw this, didn't see the comment. Could the LLVM IR function you provided look good as a test? I assume we want to exercize `isAArch64FrameOffsetLegal` here.

**Comment by fhahn** [comment]
> @fhahn, sorry, just saw this, didn't see the comment. Could the LLVM IR function you provided look good as a test? I assume we want to exercize `isAArch64FrameOffsetLegal` here.

Yes, it exposes at least the crash. Might be good to check if there are any other test gaps.

**Comment by fhahn** [comment]
@antoniofrighetto did you get a chance to add the test?

**Comment by antoniofrighetto** [comment]
@fhahn, sorry for the delay in getting back, and thanks for the ping; this took a bit more than expected. I pushed the test (and variants thereof) in 88d0ceb68936d8905bfc1724a18690b3b6115f94. This seems to reveal that we fail in emitting a `stlurb` and `stlurh` when leveraging GISel ([here](https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/AArch64/Atomics/aarch64-atomic-store-rcpc_immo.ll#L360)). I think https://github.com/antoniofrighetto/llvm-project/commit/f253bfc95eda2fc48e0bef2ed1f49db1ed63dbf2 could solve the missed opportunity, but I’m not sure why we bail out from `selectImpl` this time, for these two instructions. Hope I can take a look at this soon.

**Pull request merged**